test/InputOutput/hybrid2dbox.jl

```diff
@@ -116,4 +116,5 @@ end
     restart_Y = InputOutput.read_field(reader, "Y") # read fieldvector from hdf5 file
     close(reader)
     @test restart_Y == Y # test if restart is exact
+    @test typeof(Y) == typeof(restart_Y)
```

Comment:
I think these tests are actually redundant in that `Y == restart_Y` effectively already tests for type diagonalization through `Fields.rcompare`. Should we remove them?

Reply:
I found this to be necessary. The test for `Y == Y_restarted` for deep atmosphere passes even without fixing the bug. I didn't investigate why `rcompare` doesn't catch that

Reply:
Hm, but the method definition for `rcompare` is

```julia
rcompare(x::T, y::T) where {T <: Union{FieldVector, NamedTuple}} =
    _rcompare(true, x, y)
```
So, if the types aren't exactly equal, it will method error. Are `restart_Y` and `Y` `FieldVector`s or something else?

Reply:
Here's a reproducer
```

Reply:
This is why I added the explicit type comparison, but I am happy to remove it if you figure out what's going wrong with `rcompare`

Reply:
Ah, I know what's wrong:

```julia
julia> Fields.rcompare(restart_Y, Y)
ERROR: MethodError: no method matching rcompare(::ClimaCore.Fields.FieldVector{…}, ::ClimaCore.Fields.FieldVector{…})

Closest candidates are:
  rcompare(::T, ::T) where T<:Union{NamedTuple, ClimaCore.Fields.FieldVector}
   @ ClimaCore ~/repos/ClimaCore.jl/src/Fields/fieldvector.jl:456

Stacktrace:
 [1] top-level scope
   @ REPL[41]:1
Some type information was truncated. Use `show(err)` to see complete types.

julia> restart_Y == Y
true
```

Reply:
```
julia> @which (==)(restart_Y,Y)
==(A::AbstractArray, B::AbstractArray)
     @ Base abstractarray.jl:2937
```

Reply:
Thanks! Let me try this out (we'd better be sure that `rcompare` is working correctly if we're using it for correctness tests!)

Reply:
Ahhhh, it's _not_ calling `rcompare`! The types don't match, and so it falls back on `==` for abstract arrays. Perhaps we should catch the case when it doesn't, and throw an error? Good catch!

Reply:
Calling `rcompare` directly will work, but I think we'd better just define a version that breaks when comparing two different types